Documentation/op-guide/grpc_proxy.md

```diff
@@ -0,0 +1,49 @@
+# gRPC proxy
+
+*This is a pre-alpha feature, we are looking for early feedback.*
+
+gRPC proxy is a stateless etcd reverse proxy operating at gRPC layer (L7). gRPC proxy is designed to make watch and lease API horizontally scalable. The gRPC proxy can also help protect the core etcd cluster from abusive clients.
+
+The gRPC proxy supports multiple etcd server endpoints. When the proxy starts, it randomly picks one etcd server endpoint to use. This endpoint serves all requests until the proxy detects a network failure. If the gRPC proxy detects an endpoint failure, it switches to a different endpoint, if available, to hide failures from its clients. Other retry policies, such as weighted round-robin, may be supported in the future.
+
+## Scalable watch API
+
+The gRPC proxy is able to coalesce multiple client watchers (`c-watchers`) on the same key or range, and create one watcher (`s-watcher`) to the etcd server. It broadcasts the received event from the `s-watcher` to all `c-watchers`.
+
+Assuming N clients watch the same key, one gRPC proxy can reduce the watch load on the etcd server from N to 1. Users can deploy multiple gRPC proxies to further distribute server load.
+
+In the following example, three clients watch on key A. The gRPC proxy coalesces the three watchers, creating a single  watcher attached to the etcd server.
+
+```
+            +-------------+
+            | etcd server |
+            +------+------+
+                   ^ watch key A
+                   |
+           +-------+-----+
+           | gRPC proxy  | <-------+
+           |             |         |
+           ++-----+------+         |watch key A
+watch key A ^     ^ watch key A    |
+            |     |                |
+    +-------+-+  ++--------+  +----+----+
+    |  client |  |  client |  |  client |
+    |         |  |         |  |         |
+    +---------+  +---------+  +---------+
+```
+
+### Limitations
+
+To effectively coalesce multiple client watchers into a single watcher, the gRPC proxy coalesces new client watchers into an existing watcher when possible. This coalesced watcher may be out of sync with the etcd server due to network delays or buffered undelivered events. When the watch revision is unspecified, the gRPC proxy will not guarantee the client watcher will start watching from the most recent store revision. For example, if the client watches from an etcd server with revision 1000, that watcher will begin at revision 1000. If the client watches from the gRPC proxy, may begin watching from revision 990.
```

Comment:
Should this use the c-watcher / s-watcher terminology?
`
To effectively coalesce multiple client watchers into a single watcher, the gRPC proxy coalesces new c-watchers into an existing s-watcher when possible. This coalesced s-watcher may be out of sync with the etcd server due to network delays or buffered undelivered events. When the watch revision is unspecified, the gRPC proxy will not guarantee the c-watcher will start watching from the most recent store revision. For example, if a client watches from an etcd server with revision 1000, that watcher will begin at revision 1000. If a client watches from the gRPC proxy, may begin watching from revision 990.`